include/seqan3/core/bit_manipulation.hpp

```diff
@@ -0,0 +1,89 @@
+// ============================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ============================================================================
+//
+// Copyright (c) 2006-2018, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2018, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides utility functions for bit twiddling.
+ * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
+ * \sa https://en.wikipedia.org/wiki/Bit_manipulation
+ */
+
+#pragma once
+
+#include <meta/meta.hpp>
+
+#include <utility>
+
+#include <seqan3/core/concept/tuple.hpp>
+#include <seqan3/core/pod_tuple.hpp>
+#include <seqan3/core/type_list.hpp>
+#include <seqan3/core/metafunction/basic.hpp>
+#include <seqan3/core/metafunction/template_inspection.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\brief Is this number a power of two.
+ * \ingroup core
+ *
+ * \param[in] n The number to check.
+ *
+ * \returns True if *n* is a power of two. False otherwise.
+ *
+ * \sa https://graphics.stanford.edu/~seander/bithacks.html#DetermineIfPowerOf2
+ */
+constexpr bool is_power_of_two(size_t n)
+{
+    return n > 0 && (n & (n-1)) == 0;
```

Comment:
isn't (popcnt == 1) faster?

Reply:
Can you do a small benchmark? :D

Reply:
I'm not sure if popcnt is possible within constexpr contexts.

But I can add those functions if you want!

Reply:
As part of this PR or extra?

Reply:
Without `-march=native`:

```
------------------------------------------------------------------------------------
Benchmark                                             Time           CPU Iterations
------------------------------------------------------------------------------------
is_power_of_two_popcount<unsigned>                    9 ns          9 ns   74986854
is_power_of_two_popcount<unsigned long>               9 ns          9 ns   81160711
is_power_of_two_popcount<unsigned long long>          8 ns          8 ns   82845074
is_power_of_two_arithmetic                            7 ns          7 ns  103822138
is_power_of_two_seqan3                                7 ns          7 ns  103077653
```

with `-march=native`:


```
is_power_of_two_popcount<unsigned>                    7 ns          7 ns  103030113
is_power_of_two_popcount<unsigned long>               7 ns          7 ns  104128313
is_power_of_two_popcount<unsigned long long>          7 ns          7 ns  106325271
is_power_of_two_arithmetic                            7 ns          7 ns  100934860
is_power_of_two_seqan3                                7 ns          7 ns  101724033
```

I think in this case it is not so worthwhile.